File: 2023-03/python/part2.py
import re

NUMBER_REGEX = re.compile(r"(\d+)")
STAR_REGEX = re.compile(r"\*")


def get_range(vec, imin, imax, inclusive=False):
    return vec[max(0, imin): min(len(vec) - 1, imax) + 1 if inclusive else min(len(vec), imax)]


def adjacent_numbers(lines, line_idx, col_idx):
    return sum(
        ([int(c.group()) for c in NUMBER_REGEX.finditer(line) if c.start() - 1 <= col_idx < c.end() + 1]
         for line in get_range(lines, line_idx - 1, line_idx + 1, inclusive=True)), [])


def run(file):
    lines = file.readlines()
    result = 0
    for i, line in enumerate(lines):
        for c in STAR_REGEX.finditer(line):
            numbers = adjacent_numbers(lines, i, c.start())
            if len(numbers) == 2:
                result += numbers[0] * numbers[1]
    return result


if __name__ == "__main__":
    with open("../input.txt", 'r') as f:
        print(run(f))

File: 2023-03/rust/part2.rs
pub mod get_relaxed;
use crate::get_relaxed::GetRelaxed;

use lazy_static::lazy_static;
use regex::Regex;
use std::fs::read_to_string;

lazy_static! {
    static ref NUMBER_REGEX: Regex = Regex::new(r"\d+").unwrap();
    static ref STAR_REGEX: Regex = Regex::new(r"\*").unwrap();
}

fn adjacent_numbers(lines: &Vec<&str>, line_idx: usize, col_idx: usize) -> Vec<i32> {
    lines
        .get_relaxed(line_idx as i32 - 1..=line_idx as i32 + 1)
        .iter()
        .map(|line| {
            NUMBER_REGEX
                .captures_iter(line)
                .map(|c| c.get(0).unwrap())
                .filter(|m| {
                    m.range().start as i32 - 1 <= col_idx as i32 && col_idx < m.range().end + 1
                })
                .map(|m| m.as_str().parse::<i32>().unwrap())
        })
        .flatten()
        .collect()
}

fn run(filename: &str) -> i32 {
    let file_content = read_to_string(filename).unwrap();
    let lines: Vec<&str> = file_content.lines().collect();
    lines
        .iter()
        .enumerate()
        .map(|(line_idx, line)| {
            STAR_REGEX
                .captures_iter(line)
                .map(|c| {
                    let m = c.get(0).unwrap();
                    adjacent_numbers(&lines, line_idx, m.range().start)
                })
                .filter(|numbers| numbers.len() == 2)
                .map(|numbers| numbers[0] * numbers[1])
                .sum::<i32>()
        })
        .sum()
}

fn main() {
    let result = run("../input.txt");
    println!("{result}");
}
